### PR TITLE
[merged] Retrieve image data from DBus API and modify diff and scan features.

### PR DESCRIPTION
**Pull request by AmartC**

### CODE REVIEW

File: Atomic/atomic.py [hunk 1]
```diff
@@ -731,14 +734,24 @@ def get_col_lengths(_images):
...
+                image_dict["created"] = created
+                image_dict["virtualsize"] = virtual_size
+                image_dict["type"] = image_type
+                image_dict["vuln_info"] = json.loads(all_vuln_info)[image["Id"]]
+                all_image_info.append(image_dict)
+                if self.useTTY:
+                    util.write_out(col_out.format(indicator, repo, tag, image_id, created, virtual_size, image_type))
+            if not self.useTTY:
+                return json.dumps(all_image_info)
             util.write_out("")
```
Comment (rhatdan) [inline Atomic/atomic.py]: I think we should always return the json.dumps.  Perhaps split out the writting to tty into a separate function sing all_image_info?

Comment (AmartC) [inline Atomic/atomic.py]: so have the sing_all_image_info function call the images function which returns the json string which is then properly parsed so we can get the information and then display it?

Comment (rhatdan) [inline Atomic/atomic.py]: Yes that is what I am thinking.

File: Atomic/atomic.py [hunk 2]
```diff
@@ -660,7 +660,7 @@ def is_dangling(self, image):
             return True
         return False
 
-    def images(self):
+    def sing_all_image_info(self):
         def split_repo_tags(_images):
```
Comment (rhatdan) [inline Atomic/atomic.py]: sing?

Comment (AmartC) [inline Atomic/atomic.py]: I'll change it to get_all_image_info

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
@rh-atomic-bot retest

**Comment by rhatdan** [comment]
@AmartC Looks like this patch breaks system containers.

File: Atomic/atomic.py [hunk 3]
```diff
@@ -1050,6 +1078,17 @@ def set_debug(self):
         if self.args.debug:
             self.debug = True
 
+    def get_all_vulnerable_info(self):
+        """
+        Will simply read and return the entire /var/lib/atomic/scan_summary.json as a JSON string so it can then be parsed appropriately.
+        """
+        try:
+            return json.dumps(json.loads(open(os.path.join(self.results, "scan_summary.json"), "r").read()))
+            #return json.dumps(summary_results)
```
Comment (giuseppe) [inline Atomic/atomic.py]: we do we need to `loads` and then `dumps`?  Is it enough to use what `.read()` returns?

Also the commented code should be removed.

Comment (AmartC) [inline Atomic/atomic.py]: i removed the commented code. also yeah when i had tried it in the past it didn't seem to work but i gave another attempt and now it looks good. i'll implement those changes.

Comment (giuseppe) [inline Atomic/atomic.py]: can you verify both Python2 and Python3?  There might be some encoding issues with Python3 that you don't catch when testing only with Python2.

Comment (AmartC) [inline Atomic/atomic.py]: okay, how should i do that?

File: Atomic/atomic.py [hunk 4]
```diff
@@ -711,7 +709,27 @@ def get_col_lengths(_images):
...
                                               "TYPE"))
+
+            for image in image_data:
+                util.write_out(col_out.format(image['indicator'], image['repo'], image['tag'], image['id'], image['created'], image['virtual_size'], image['type']))
+            util.write_out("")
+
+
+
+
+
```
Comment (giuseppe) [inline Atomic/atomic.py]: please leave only one empty line here

File: Atomic/atomic.py [hunk 5]
```diff
@@ -1050,6 +1078,17 @@ def set_debug(self):
...
 
+    def get_all_vulnerable_info(self):
+        """
+        Will simply read and return the entire /var/lib/atomic/scan_summary.json as a JSON string so it can then be parsed appropriately.
+        """
+        try:
+            return json.dumps(json.loads(open(os.path.join(self.results, "scan_summary.json"), "r").read()))
+            #return json.dumps(summary_results)
+        except IOError:
+            return ""
```
Comment (giuseppe) [inline Atomic/atomic.py]: should be `return "{}"`, the empty string is not a valid JSON document.

File: Atomic/atomic.py [hunk 6]
```diff
@@ -711,7 +709,27 @@ def get_col_lengths(_images):
...
+
+
+
+
+
+    def images(self):
+        _images = self.get_images()
+
+        used_image_ids = [x['ImageID'] for x in self.get_containers()]
+        all_image_info = [_images]
```
Comment (giuseppe) [inline Atomic/atomic.py]: why do we need this?  IIUIC the element in the first position is different than the rest?

Comment (AmartC) [inline Atomic/atomic.py]: Yes, it's used by the function to output the images

File: Atomic/atomic.py [hunk 7]
```diff
@@ -711,7 +709,27 @@ def get_col_lengths(_images):
                                               "CREATED",
                                               "VIRTUAL SIZE",
                                               "TYPE"))
+
+            for image in image_data:
```
Comment (giuseppe) [inline Atomic/atomic.py]: we should change it so that we can iterate over `_images` instead of having `image_data`. 

It will also simplify the `get_col_lengths` function, that can do: `repo_tags = [[i["repo"], i["tag"]] for i in _images]` and get rid completely of `split_repo_tags`.

**Comment by giuseppe** [comment]
while reviewing it, I've done some changes to try it out.

Please let me know what you think of them and feel free to amend these changes in your version.  

Tests should pass as well.

```
diff --git a/Atomic/atomic.py b/Atomic/atomic.py
index 8d32502..a8afe10 100644
--- a/Atomic/atomic.py
+++ b/Atomic/atomic.py
@@ -662,15 +662,6 @@ class Atomic(object):
         return False
 
     def display_all_image_info(self):
-        def split_repo_tags(_images):
-            sub_list = [item.split(":") for sublist in _images for item
-                         in sublist['RepoTags']]
-            repo_tags = []
-            for repo in sub_list:
-                if len(repo) > 2:
-                    repo = [repo[0] + repo[1], repo[2]]
-                repo_tags.append(repo)
-            return repo_tags
 
         def get_col_lengths(_images):
             '''
@@ -679,7 +670,7 @@ class Atomic(object):
             :return: a set with len of repository and tag
             If there are no images, return 1, 1
             '''
-            repo_tags = split_repo_tags(_images)
+            repo_tags = [[i["repo"], i["tag"]] for i in _images]
             # Integer additions below are for column padding
             # 7 == 1 for dangling, 2 for spacing, 4 for highlighting
             if repo_tags:
@@ -688,10 +679,8 @@ class Atomic(object):
             else:
                 return 1, 1
 
-        all_images_info = json.loads(self.images())
-        _images = all_images_info[0]
+        _images = self.images()
         if len(_images) >= 0:
-            image_data = all_images_info[1:]
             _max_repo, _max_tag = get_col_lengths(_images)
             if self.args.truncate:
                 _max_id = 14
@@ -710,25 +699,23 @@ class Atomic(object):
                                               "VIRTUAL SIZE",
                                               "TYPE"))
 
-            for image in image_data:
-                util.write_out(col_out.format(image['indicator'], image['repo'], image['tag'], image['id'], image['created'], image['virtual_size'], image['type']))
+            for image in _images:
+                util.write_out(col_out.format(image['indicator'], image['repo'], image['tag'],
+                                              image['id'], image['created'], image['virtual_size'],
+                                              image['type']))
             util.write_out("")
 
-
-
-
-
     def images(self):
         _images = self.get_images()
 
         used_image_ids = [x['ImageID'] for x in self.get_containers()]
-        all_image_info = [_images]
+        all_image_info = []
 
         if len(_images) >= 0:
             vuln_ids = self.get_vulnerable_ids()
             all_vuln_info = json.loads(self.get_all_vulnerable_info())
 
-            for image in self.get_images():
+            for image in _images:
                 image_dict = dict()
                 repo, tag = image["RepoTags"][0].rsplit(":", 1)
                 if "Created" in image:
@@ -769,7 +756,7 @@ class Atomic(object):
                 image_dict["indicator"] = indicator
                 all_image_info.append(image_dict)
 
-            return json.dumps(all_image_info)
+        return all_image_info
 
     def _check_if_image_present(self):
         self.inspect = self._inspect_image()
@@ -1083,10 +1070,9 @@ class Atomic(object):
         Will simply read and return the entire /var/lib/atomic/scan_summary.json as a JSON string so it can then be parsed appropriately.
         """
         try:
-            return json.dumps(json.loads(open(os.path.join(self.results, "scan_summary.json"), "r").read()))
-            #return json.dumps(summary_results)
+            return open(os.path.join(self.results, "scan_summary.json"), "r").read()
         except IOError:
-            return ""
+            return "{}"
 
 
     def get_vulnerable_ids(self):
diff --git a/atomic_dbus.py b/atomic_dbus.py
index b14c388..d2fb61e 100755
--- a/atomic_dbus.py
+++ b/atomic_dbus.py
@@ -253,7 +253,7 @@ class atomic_dbus(slip.dbus.service.Object):
     def Images(self):
         args = self.Args()
         self.atomic.set_args(args)
-        return self.atomic.images()
+        return json.dumps(self.atomic.images())
 
     # The Vulnerable method will send back information that says whether or not an installed container image is vulnerable
     @slip.dbus.polkit.require_auth("org.atomic.read")
```

File: Atomic/atomic.py [hunk 8]
```diff
@@ -711,7 +697,21 @@ def get_col_lengths(_images):
                                               "CREATED",
                                               "VIRTUAL SIZE",
                                               "TYPE"))
-            for image in self.get_images():
+
+            for image in _images:
+                util.write_out(col_out.format(image['indicator'], image['repo'], image['tag'], image['id'], image['created'], image['virtual_size'], image['type']))
```
Comment (giuseppe) [inline Atomic/atomic.py]: please split this line

File: Atomic/atomic.py [hunk 9]
```diff
@@ -1050,6 +1059,16 @@ def set_debug(self):
         if self.args.debug:
             self.debug = True
 
+    def get_all_vulnerable_info(self):
+        """
+        Will simply read and return the entire /var/lib/atomic/scan_summary.json as a JSON string so it can then be parsed appropriately.
```
Comment (giuseppe) [inline Atomic/atomic.py]: this can be splitted too

File: atomic_dbus.py [hunk 10]
```diff
@@ -244,6 +247,22 @@ def Update(self, image):
...
 
+    # The Images method will list all installed container images on the system.
+    @slip.dbus.polkit.require_auth("org.atomic.read")
+    @dbus.service.method("org.atomic", in_signature='', out_signature='s')
+    def Images(self):
+        args = self.Args()
+        self.atomic.set_args(args)
+        return json.dumps(self.atomic.images())
+
+    # The Vulnerable method will send back information that says whether or not an installed container image is vulnerable
```
Comment (giuseppe) [inline atomic_dbus.py]: and this one

**Comment by giuseppe** [comment]
just a few minor things and please change the commit message to reflect the changes in the patch and not all the updates to it.  After that, I think it is ready to go

File: Atomic/atomic.py [hunk 11]
```diff
@@ -97,7 +97,7 @@ class Atomic(object):
                 "${IMAGE}"]
 
     results = '/var/lib/atomic'
-    skull = (u"\u2620").encode('utf-8')
+    skull = (u"\u2620")
```
Comment (giuseppe) [inline Atomic/atomic.py]: this kind of changes should go in a separate patch

**Comment by giuseppe** [comment]
the commit message has still some duplicate lines, they should be removed and in general describe the current status of the patch, not the changes done to it

File: Atomic/atomic.py [hunk 12]
```diff
@@ -747,36 +769,34 @@ def get_col_lengths(_images):
...
-                    indicator = "*"
+                    image_dict["is_dangling"] = True
                 elif image['Id'] in used_image_ids:
-                    indicator = ">"
+                    image_dict["is_dangling"] = False
+                    image_dict["used_image"] = True
                 else:
-                    indicator = ""
+                    image_dict["is_dangling"] = False
+                    image_dict["used_image"] = False
```
Comment (rhatdan) [inline Atomic/atomic.py]: How about
```
+              image_dict["is_dangling"] = False
+               image_dict["used_image"] = False
                 if self.is_dangling(repo):
-                    indicator = "*"
+                    image_dict["is_dangling"] = True
                 elif image['Id'] in used_image_ids:
-                    indicator = ">"
+                    image_dict["is_dangling"] = False
-                 else:
-                    indicator = ""
```

File: Atomic/atomic.py [hunk 13]
```diff
@@ -747,36 +769,34 @@ def get_col_lengths(_images):
...
 
                 if image['Id'] in vuln_ids:
-                    space = " " if len(indicator) < 1 else ""
-                    indicator = indicator + self.skull + space
+                    image_dict["vulnerable"] = True
+                else:
+                    image_dict["vulnerable"] = False
```
Comment (rhatdan) [inline Atomic/atomic.py]: image_dict["vulnerable"]  = image['Id'] in vuln_ids

File: Atomic/atomic.py [hunk 14]
```diff
@@ -747,36 +769,34 @@ def get_col_lengths(_images):
...
-                if self.args.quiet:
-                    util.write_out(image_id)
+                image_dict["repo"] = repo
+                image_dict["tag"] = tag
+                image_dict["id"] = image_id
+                image_dict["created"] = created
+                image_dict["virtual_size"] = virtual_size
+                image_dict["type"] = image_type
+                if image["Id"] in all_vuln_info.keys():
+                    image_dict["vuln_info"] = all_vuln_info[image["Id"]]
```
Comment (rhatdan) [inline Atomic/atomic.py]: if image_dict["vulnerable"]:
  image_dict["vuln_info"] = all_vuln_info[image["Id"]]

**Comment by rhatdan** [comment]
@rh-atomic-bot r+
